### PR TITLE
Unicode scripts pretokenizer

### PR DESCRIPTION
**Pull request by Narsil**
Adding a new PreTokenizer for UnicodeScripts.
    
    The core idea is to detect the "Script" of every char in a String
    and split whenever there is a change of Script.
    
    For instance
    
    "どこで生れ。Yes" -> "どこで生れ", "。", "Yes"
    
    The first chars are all considered `Han`, "。" is `Common` and the rest are
    `Latin`.
    This PreTokenizer will be helpful to disentangle pre-processing done in
    `spm` and be able to reuse it better when needed.

### CODE REVIEW

File: tokenizers/src/pre_tokenizers/unicode_scripts/pre_tokenizer.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            Script::Katakana => Script::Han,
+            script => script,
+        }
+    }
+}
+
+impl PreTokenizer for UnicodeScripts {
+    fn pre_tokenize(&self, pretokenized: &mut PreTokenizedString) -> Result<()> {
+        pretokenized.split(|_, normalized| {
+            let mut last_script = Script::Any;
```
Comment (n1t0) [inline tokenizers/src/pre_tokenizers/unicode_scripts/pre_tokenizer.rs]: By doing so, are you sure that the offset 0 will always be the first in `ranges`?

File: tokenizers/src/pre_tokenizers/unicode_scripts/pre_tokenizer.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
+use serde::{Deserialize, Serialize};
+
+use crate::pre_tokenizers::unicode_scripts::scripts::{get_script, Script};
+use crate::tokenizer::{normalizer::Range, PreTokenizedString, PreTokenizer, Result};
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(tag = "type")]
+pub struct UnicodeScripts {}
```
Comment (n1t0) [inline tokenizers/src/pre_tokenizers/unicode_scripts/pre_tokenizer.rs]: Did you check that the serialization works as expected with an empty C-Style struct? (cf https://github.com/huggingface/tokenizers/blob/master/tokenizers/src/pre_tokenizers/whitespace.rs where we `impl Deserialize` manually, and it seems similar). Maybe we want to use a unit-struct for consistency reasons if we don't have any field here.

File: tokenizers/src/pre_tokenizers/unicode_scripts/unicode.rs [hunk 3]
```diff
@@ -0,0 +1,2093 @@
+// Generated by modified Perl script at https://github.com/google/sentencepiece/blob/master/data/gen_unicode_scripts_code.pl
```
Comment (n1t0) [inline tokenizers/src/pre_tokenizers/unicode_scripts/unicode.rs]: This file seems to be a duplicate of `unicode_scripts/scripts.rs`

File: tokenizers/src/pre_tokenizers/unicode_scripts/pre_tokenizer.rs (hunk 1 above)
Comment (Narsil) [reply tokenizers/src/pre_tokenizers/unicode_scripts/pre_tokenizer.rs]: I am but I am going to go in favor of an Option for last_script. I guess that will be clearer to the reader.

File: tokenizers/src/pre_tokenizers/unicode_scripts/pre_tokenizer.rs (hunk 2 above)
Comment (Narsil) [reply tokenizers/src/pre_tokenizers/unicode_scripts/pre_tokenizer.rs]: Will do

File: tokenizers/src/pre_tokenizers/unicode_scripts/unicode.rs (hunk 3 above)
Comment (Narsil) [reply tokenizers/src/pre_tokenizers/unicode_scripts/unicode.rs]: Yes, I kinda messed up my rebases

**Pull request closed without merging**